### PR TITLE
Sidecar container spec for druid

### PR DESCRIPTION
**Pull request by cintoSunny**
Fixes https://github.com/druid-io/druid-operator/issues/275

### Description

Currently, the Druid operator does not allow the addition of a sidecar container to Druid deployments. The only way is to inject sidecars by mutation webhooks like istio. But many orgs may not allow setting up istio or it could be an additional effort to add them. 

The sidecar containers can be added like below:

```
sidecarContainer:
- configs:
    imageName: docker.com/ns/samplesimage:latest
    containerName: sidecar
    command: /bin/start.sh
    pullPolicy: Always
...
```

This PR has:
- [x] been tested on a real K8S cluster to ensure creation of a sidecar container.
- [x] been tested for backward compatibility on a real K*S cluster by applying the changes introduced here on an existing Druid cluster. If there are any backward incompatible changes then they have been noted in the PR description.
- [ ] added documentation for new or modified features or behaviors.

<hr>

##### Key changed/added files in this PR
 * `handler.go`
 * `druid_types.go`
 * `handler_test.go`
 * `broker-statefulset-sidecar.yaml`
 * `druid-test-cr-sidecar.yaml`
 *  `druid.apache.org_druids.yaml`

### CODE REVIEW

File: controllers/druid/handler.go [hunk 1]
```diff
@@ -1302,30 +1302,58 @@ func makePodTemplate(nodeSpec *v1alpha1.DruidNodeSpec, m *v1alpha1.Druid, ls map
...
+			containers = append(containers,
+				v1.Container{
+					Image:           configs["imageName"],
+					Name:            configs["containerName"],
+					Resources:       resource,
+					VolumeMounts:    getVolumeMounts(nodeSpec, m),
+					Command:         []string{configs["command"]},
+					Args:            args,
+					ImagePullPolicy: v1.PullPolicy(configs["pullPolicy"]),
+					SecurityContext: &v1.SecurityContext{
```
Comment (AdheipSingh) [inline controllers/druid/handler.go:1342]: we can leave for the user to add securityContext, this is hardcoding the User fields.
we can keep this as we do for druid nodes https://github.com/druid-io/druid-operator/blob/master/apis/druid/v1alpha1/druid_types.go#L87

File: apis/druid/v1alpha1/druid_types.go [hunk 2]
```diff
@@ -24,6 +24,16 @@ import (
 // EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
 // NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
 
+// AdditionalContainer defines the additional sidecar container
+type AdditionalContainer struct {
+	// List of configurations to use which are not present or to override default implementation configurations
+	// +optional
+	Configs   map[string]string       `json:"configs,omitempty"`
+	Args      []string                `json:"args,omitempty"`
+	User      int64                   `json:"user,omitempty"`
```
Comment (AdheipSingh) [inline apis/druid/v1alpha1/druid_types.go:33]: let keep it to container security context, let users add the specs.

File: controllers/druid/handler.go [hunk 3]
```diff
@@ -1302,30 +1302,58 @@ func makePodTemplate(nodeSpec *v1alpha1.DruidNodeSpec, m *v1alpha1.Druid, ls map
...
+	if m.Spec.SidecarContainer != nil {
+
+		for _, containerList := range m.Spec.SidecarContainer {
+			configs := containerList.Configs
+			args := containerList.Args
+			resource := containerList.Resources
+
+			containers = append(containers,
+				v1.Container{
+					Image:           configs["imageName"],
```
Comment (AdheipSingh) [inline controllers/druid/handler.go:1335]: can we add a constants blob with respect to all the strings, something like
```
const (
  ImageName = "imageName"
  
)
```
just to make it cleaner. you can add it in interface.go.

File: apis/druid/v1alpha1/druid_types.go [hunk 4]
```diff
@@ -132,6 +142,9 @@ type DruidSpec struct {
 	// that is, it must match regex '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'
 	Nodes map[string]DruidNodeSpec `json:"nodes"`
 
+	// Operator deploys the sidecar container based on these properties. Sidecar will be deployed for all the Druid pods.
+	SidecarContainer []AdditionalContainer `json:"sidecarContainer,omitempty"`
```
[the remark this answers is not in the document]
Comment (cintoSunny) [reply apis/druid/v1alpha1/druid_types.go:146]: Do you think we will have a use case for sidecar in specific nodes? We can do that, but will it be overkill?

File: controllers/druid/handler.go [hunk 5]
```diff
@@ -1302,30 +1302,58 @@ func makePodTemplate(nodeSpec *v1alpha1.DruidNodeSpec, m *v1alpha1.Druid, ls map
...
+					Name:            configs["containerName"],
+					Resources:       resource,
+					VolumeMounts:    getVolumeMounts(nodeSpec, m),
```
[the remark this answers is not in the document]
Comment (cintoSunny) [reply controllers/druid/handler.go:1338]: You are right. Currently, this PR does not support volume mounts specifically for sidecar. It mounts from druid node containers. Will we need one specifically for the sidecar? I initially thought it would be okay to mount the same volume mounts. Let me know

File: controllers/druid/handler.go (hunk 1 above)
Comment (cintoSunny) [reply controllers/druid/handler.go:1342]: makes sense

File: apis/druid/v1alpha1/druid_types.go (hunk 2 above)
Comment (cintoSunny) [reply apis/druid/v1alpha1/druid_types.go:33]: i agree

File: apis/druid/v1alpha1/druid_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (AdheipSingh) [reply apis/druid/v1alpha1/druid_types.go:146]: Agreed i personally don't have a usecase or can think of one. One of the reasons to ask this was that i see this as an additional container right ? that can be any container for any of the druid nodes. 
Also how about we keep this too
```
	AdditionalContainer []AdditionalContainer `json:"additionalContainer,omitempty"`
```
I see now this PR can work for N number of use cases :), be it sidecar specific or just any other containee.
Thoughts ?

File: apis/druid/v1alpha1/druid_types.go [hunk 6]
```diff
@@ -24,6 +24,16 @@ import (
 // EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
 // NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
 
+// AdditionalContainer defines the additional sidecar container
+type AdditionalContainer struct {
+	// List of configurations to use which are not present or to override default implementation configurations
+	// +optional
+	Configs   map[string]string       `json:"configs,omitempty"`
```
[the remark this answers is not in the document]
Comment (AdheipSingh) [reply apis/druid/v1alpha1/druid_types.go:31]: yeah, ideally we should have a struct which has podConfigs and containerConfigs and that can be type embedded in the nodeSpec, ClusterSpec and ContainerSpec. That would need a lot of code change.
Lets keep it to minimal for now. That should work IMO

File: controllers/druid/handler.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (AdheipSingh) [reply controllers/druid/handler.go:1338]: since its a different container altogether lets keep the abstractions around any of the configs.

File: apis/druid/v1alpha1/druid_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (cintoSunny) [reply apis/druid/v1alpha1/druid_types.go:146]: Yes, it can. I will rename the `SidecarContainer` to `additionalContainer`. Makes more sense

File: apis/druid/v1alpha1/druid_types.go [hunk 7]
```diff
@@ -24,6 +24,22 @@ import (
...
+	Image 					 string `json:"image,omitempty"`
+	ContainerName			 string `json:"containerName,omitempty"`
+	Command					 []string `json:"command,omitempty"`
+	ImagePullPolicy 		 v1.PullPolicy `json:"imagePullPolicy,omitempty"`
+	Args                     []string                `json:"args,omitempty"`
+	ContainerSecurityContext *v1.SecurityContext     `json:"securityContext,omitempty"`
+	Resources                v1.ResourceRequirements `json:"resources,omitempty"`
+	VolumeMounts         	 []v1.VolumeMount           `json:"volumeMounts,omitempty"`
+	Env 					 []v1.EnvVar `json:"env,omitempty"`
+}
```
Comment (AdheipSingh) [inline apis/druid/v1alpha1/druid_types.go:41]: can we format this and remove //configs as a comment

File: controllers/druid/handler.go [hunk 8]
```diff
@@ -1302,30 +1302,54 @@ func makePodTemplate(nodeSpec *v1alpha1.DruidNodeSpec, m *v1alpha1.Druid, ls map
...
+					Command:         containerList.Command,
+					Args:            containerList.Args,
+					ImagePullPolicy: containerList.ImagePullPolicy,
+					SecurityContext: containerList.ContainerSecurityContext,
+					Env: 			 containerList.Env,
+				},
+			)
+		}
+	}
+
```
Comment (AdheipSingh) [inline controllers/druid/handler.go:1345]: This is absolute reference to pass in the value, in case the field is missing i expect the program to crash, ill suggest to test out the scenario and use pointers to reference, else writing functions and handling nil would be repetitive code.

**Review by AdheipSingh — CHANGES_REQUESTED** [review]

File: apis/druid/v1alpha1/druid_types.go (hunk 7 above)
Comment (cintoSunny) [reply apis/druid/v1alpha1/druid_types.go:41]: Sure. Will do that

File: controllers/druid/handler.go (hunk 8 above)
Comment (cintoSunny) [reply controllers/druid/handler.go:1345]: The current code is similar to how ports and resources are defined in the parent container
```
Ports:           nodeSpec.Ports,
Resources:       nodeSpec.Resources,
```

Comment (cintoSunny) [reply controllers/druid/handler.go:1345]: @AdheipSingh : Let me know what you think.

Comment (AdheipSingh) [reply controllers/druid/handler.go:1345]: ill run locally and confirm the changes sometime by end of the week.
Thanks

File: apis/druid/v1alpha1/druid_types.go [hunk 9]
```diff
@@ -132,6 +148,9 @@ type DruidSpec struct {
 	// that is, it must match regex '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'
 	Nodes map[string]DruidNodeSpec `json:"nodes"`
 
+	// Operator deploys the sidecar container based on these properties. Sidecar will be deployed for all the Druid pods.
+	AdditionalContainer []AdditionalContainer `json:"additionalContainer,omitempty"`
+
```
Comment (AdheipSingh) [inline apis/druid/v1alpha1/druid_types.go:153]: this is scoped at cluster scope only. so its right to assume that additional container will be common to all the nodes. IMHO its fine.

**Review by AdheipSingh — CHANGES_REQUESTED** [review]

File: apis/druid/v1alpha1/druid_types.go [hunk 10]
```diff
@@ -24,6 +24,22 @@ import (
...
+	ContainerName			 string `json:"containerName,omitempty"`
+	Command					 []string `json:"command,omitempty"`
+	ImagePullPolicy 		 v1.PullPolicy `json:"imagePullPolicy,omitempty"`
+	Args                     []string                `json:"args,omitempty"`
+	ContainerSecurityContext *v1.SecurityContext     `json:"securityContext,omitempty"`
+	Resources                v1.ResourceRequirements `json:"resources,omitempty"`
+	VolumeMounts         	 []v1.VolumeMount           `json:"volumeMounts,omitempty"`
+	Env 					 []v1.EnvVar `json:"env,omitempty"`
+}
+
```
[the remark this answers is not in the document]
Comment (AdheipSingh) [reply apis/druid/v1alpha1/druid_types.go:42]: also remove //Configs as comment  and+optional. onlywhen the type is embedded we can +optional

File: controllers/druid/handler.go [hunk 11]
```diff
@@ -1302,30 +1302,54 @@ func makePodTemplate(nodeSpec *v1alpha1.DruidNodeSpec, m *v1alpha1.Druid, ls map
...
+				v1.Container{
+					Image:           containerList.Image,
+					Name:            containerList.ContainerName,
+					Resources:       containerList.Resources,
+					VolumeMounts:    containerList.VolumeMounts,
+					Command:         containerList.Command,
+					Args:            containerList.Args,
+					ImagePullPolicy: containerList.ImagePullPolicy,
+					SecurityContext: containerList.ContainerSecurityContext,
+					Env: 			 containerList.Env,
```
[the remark this answers is not in the document]
Comment (cintoSunny) [reply controllers/druid/handler.go]: Done. Let me know if you see any issues

File: apis/druid/v1alpha1/druid_types.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (cintoSunny) [reply apis/druid/v1alpha1/druid_types.go]: Fixed. Let me know if you see any issues

**Comment by AdheipSingh** [comment]
@nishantmonu51 kindly approve to run CI checks

**Comment by AdheipSingh** [comment]
@cintoSunny can you add a documentation around adding this feature, would be really helpful :) and an example spec here https://github.com/druid-io/druid-operator/blob/master/docs/examples.md

**Comment by cintoSunny** [comment]
> @cintoSunny can you add a documentation around adding this feature, would be really helpful :) and an example spec here https://github.com/druid-io/druid-operator/blob/master/docs/examples.md

I have a sample spec for using the additional container. 
https://github.com/druid-io/druid-operator/blob/451c819e5d6c4e7a0458185fc41a79bf6055f4f9/controllers/druid/testdata/broker-statefulset-sidecar.yaml

Do you want me to add it to the `examples.md` you mentioned above. Let me know.

File: docs/examples.md [hunk 12]
```diff
@@ -238,3 +238,36 @@
...
+          securityContext:
+            runAsUser: 506
+          imagePullPolicy: Always
+          volumeMounts:
+            - name: logstore
+              mountPath: /logstore
+          env:
+            - name: SAMPLE_ENV
+              value: SAMPLE_VALUE
+```
```
Comment (AdheipSingh) [inline docs/examples.md:273]:
```
additionalContainer:
    - image: universalforwarder-sidekick:next
      containerName: forwarder
      command:
        - /bin/sidekick
      imagePullPolicy: Always
      securityContext:
        runAsUser: 506
      volumeMounts:
        - name: logstore
          mountPath: /logstore
      env:
        - name: SAMPLE_ENV
          value: SAMPLE_VALUE
      resources:
        requests:
          memory: "1Gi"
          cpu: "500m"
        limits:
          memory: "1Gi"
          cpu: "500m"
      args:
        - -loggingEnabled=true
        - -dataCenter=dataCenter
        - -environment=environment
        - -application=application
        - -instance=instance
        - -logFiles=logFiles
```

Comment (AdheipSingh) [reply docs/examples.md:273]: can you add a pointer stating this is scoped to cluster only.

File: controllers/druid/testdata/broker-statefulset-sidecar.yaml [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+  volumeClaimTemplates:
+    - metadata:
+        name: data-volume
+      spec:
+        accessModes:
+          - ReadWriteOnce
+        resources:
+          requests:
+            storage: 2Gi
+        storageClassName: gp2
```
Comment (AdheipSingh) [inline controllers/druid/testdata/broker-statefulset-sidecar.yaml:119]: identation

**Review by AdheipSingh — CHANGES_REQUESTED** [review]

**Comment by AdheipSingh** [comment]
I have tested this locally, its backward compatible. works with updated CRD. LGTM

Can you add a working example, in examples folder. It would be easier for users to test. I had to construct own yaml for test.
( the yamls in testdata wont work against the updated CRD, need to update thise. )

**Comment by cintoSunny** [comment]
> I have tested this locally, its backward compatible. works with updated CRD. LGTM
> 
> Can you add a working example, in examples folder. It would be easier for users to test. I had to construct own yaml for test. ( the yamls in testdata wont work against the updated CRD, need to update thise. )

Sorry about that. I had to remove some company-specific things, hence left out a few parts. 
I changed it to your working example.

**Review by AdheipSingh — APPROVED** [review]

File: docs/examples.md (hunk 12 above)
Comment (cintoSunny) [reply docs/examples.md]: I added the above snippet. The one I added had a couple of copy paste errors. sorry about that.
Also added the cluster scope text to features.md. let me know what you think

File: controllers/druid/testdata/broker-statefulset-sidecar.yaml (hunk 13 above)
Comment (cintoSunny) [reply controllers/druid/testdata/broker-statefulset-sidecar.yaml]: done
